HeartbeatLogging/Sources/HeartbeatInfo.swift

```diff
@@ -14,151 +14,108 @@
-
-    // Store the written `heartbeat` in the `heartbeatsByTimePeriodCache`.
-    latestHeartbeatByTimePeriod.store(heartbeat)
-
-    // Increment `tailIndex`, wrapping around to the start if needed.
-    tailIndex = (tailIndex + 1) % circularQueue.capacity
+  /// <#Description#>
+  /// - Parameter elements: <#elements description#>
+  init(elements: [Element]) {
+    circularQueue = elements
   }
 
-  /// Returns the last heartbeat from a given time period.
-  /// - Parameter timePeriod: Time period where a heartbeat may occur.
-  /// - Returns: Optionally, the last  `Heartbeat` to occur in the given time period.
-  func lastHeartbeat(forTimePeriod timePeriod: TimePeriod) -> Heartbeat? {
-    latestHeartbeatByTimePeriod[timePeriod]
-  }
-
-  /// A cache mapping `TimePeriod` keys to `Heartbeat` values.
-  ///
-  /// This type's API are considered to be constant time as they are bounded by the number of cases in
-  /// the `TimePeriod` enum.
-  ///
-  /// - Note: This type has reference semantics.
-  private final class HeartbeatByTimePeriodCache: Codable {
-    private lazy var cache: [TimePeriod: Heartbeat] = [:]
-
-    /// Removes a given heartbeat from the cache.
-    /// - Parameter heartbeat: The heartbeat to remove.
-    func remove(_ heartbeat: Heartbeat) {
-      // The below operation is considered constant time because the cache has
-      // a bounded number of keys (see the `TimePeriod` type).
-      cache = cache.filter { $1 /* cachedHeartbeat */ != heartbeat }
+  /// Pushes an element to the back of the buffer, returning the element that was overriten if the capacity is reached.
+  /// - Parameter element: The element to push to the back of the buffer.
+  /// - Complexity: O(1)
+  @discardableResult
+  mutating func push(_ element: Element) -> Element? {
+    guard circularQueue.capacity > 0 else {
+      // Do not append if `capacity` is less than or equal 0.
+      return nil
     }
 
-    /// Stores a given heartbeat to the cache.
-    /// - Parameter heartbeat: The heartbeat to store.
-    func store(_ heartbeat: Heartbeat) {
-      // The below operation is considered constant time because a heartbeat
-      // has a bound number of time periods (see the `TimePeriod` type).
-      heartbeat.timePeriods.forEach { cache[$0] = heartbeat }
+    defer {
+      // Increment index, wrapping around to the start if needed.
+      tailIndex += 1
+      tailIndex %= circularQueue.capacity
     }
 
-    subscript(timePeriod: TimePeriod) -> Heartbeat? {
-      cache[timePeriod]
-    }
+    let replaced = circularQueue[tailIndex]
+    circularQueue[tailIndex] = element
+    return replaced
   }
-}
 
-private extension Date {
-  /// Calculates the time interval since a given `date`.
-  func timeIntervalSince(_ date: Date) -> TimeInterval {
-    timeIntervalSinceReferenceDate - date.timeIntervalSinceReferenceDate
+  func makeIterator() -> IndexingIterator<[Element]> {
+    circularQueue
+      .compactMap { $0 } // Remove `nil` elements.
```

Comment:
Why are there nil elements?

Reply:
The `circularQueue` is an array with a fixed size (Buffer is initialized with fixed capacity at initialization). Empty/placeholder array elements are `nil`. 

```swift
let fixedCapacityArray: Int? = Array(repeating: nil, count: 3)  // [nil, nil, nil]
```